src/Oscoin/Crypto/Address.hs

```diff
@@ -0,0 +1,97 @@
+{- | An address format for the Oscoin blockchain loosely based on Parity's SS58
+-- and on BTC's Base58Check.
+--
+-- See docs/address_format.md for an extensive documentation.
+--
+--}
+
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Oscoin.Crypto.Address
+    ( Address(addressPrefix, addressPayload)
+    , AddressPrefix
+    , AddressPayload
+
+    -- * Creating an Address
+    , fromPublicKey
+
+    -- * Serialising & deserialising addresses
+    , Serialisation.serialiseAddress
+    , Serialisation.deserialiseAddress
+
+    -- * Pretty-printing an Address
+    , fmtAddress
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Address.Internal
+import           Oscoin.Crypto.Address.Serialisation as Serialisation
+import           Oscoin.Crypto.Hash
+import           Oscoin.Crypto.PubKey
+import           Oscoin.Environment
+
+import qualified Codec.Serialise as CBOR
+import           Data.ByteString.BaseN (encodeBase58, encodedText)
+import           Data.ByteString.Short (toShort)
+import           Data.Tagged
+import qualified Formatting as F
+import           Formatting.Buildable
+
+{------------------------------------------------------------------------------
+    Creating an Address
+------------------------------------------------------------------------------}
+
+-- | NOTE(adn) It's conceivable that we might want to crate a type to carry
+-- around the context needed to create an 'Address', but for now single
+-- arguments should do.
+fromPublicKey
+    :: forall c.
+       ( CBOR.Serialise (ShortHash c)
+       , Hashable c (PublicKey c)
+       )
+    => Environment
+    -- ^ The 'Environment' the node is running on.
+    -> PublicKey c
+    -- ^ A 'PublicKey'.
+    -> Address c
+fromPublicKey env pk =
+    let prefix = AddressPrefix
+          { addressFormat   = AddressFormat AddressFormatTag_V1
+          , addressType     = AddressType addressTypeTagFromEnv
+          , protocolVersion = ProtocolVersion ProtocolVersion_V1
+          }
+        payload = toShort
+                . toS
+                . CBOR.serialise @(ShortHash c)
+                . shortHash
+                $ pk
+    in Address prefix (AddressPayload . Tagged $ payload)
+  where
+    -- NOTE(adn) This is a bit iffy, as we might want to create
```

Comment:
I think this is what I *really* need, alongside the environment: https://github.com/oscoin/oscoin/pull/419/files#diff-2b4a1052f78920146b335fda52b51cd2R68